src/types/Query.ts

```diff
@@ -0,0 +1,70 @@
+import {
+  FieldPath,
+  OrderByDirection,
+  WhereFilterOp,
+} from '@firebase/firestore-types'
+import { SerializerOptions } from './Serializer'
+
+type KeyHack = string & {} // hack to also allow strings
+
+// here we get the "key" from our data, to add intellisense for any "orderBy" in the queries and such.
+export type OrderByArray<Doc extends object = {}, Key = keyof Doc> = [
+  Key | FieldPath | KeyHack,
+  OrderByDirection
+]
+export type OrderByItem<Doc extends object = {}, Key = keyof Doc> =
+  | OrderByArray<Doc>
+  | Key
+  | KeyHack
+export type OrderByType<Doc extends object = {}> =
+  | OrderByItem<Doc>
+  | OrderByArray<Doc>[]
+
+export type WhereItem<Doc extends object = {}, Key = keyof Doc> = [
+  Key | FieldPath | KeyHack,
+  WhereFilterOp,
+  unknown,
+  SerializerOptions?
+]
+export type WhereArray<Doc extends object = {}> = WhereItem<Doc>[]
+export type WhereType<Doc extends object = {}> =
+  | WhereItem<Doc>
+  | WhereArray<Doc>
+
+export type CollectionQueryType<Doc extends object = {}> = {
+  limit?: number
+  orderBy?: OrderByType<Doc>
+  where?: WhereType<Doc>
+  isCollectionGroup?: boolean
+
+  /**
+   * For now, this can only be a number, since it has to be JSON serializable.
+   *
+   * **TODO** allow DocumentSnapshot here too. This will probably be used with a useStaticCollection hook in the future.
+   */
+  startAt?: number
+  /**
+   * For now, this can only be a number, since it has to be JSON serializable.
+   *
+   * **TODO** allow DocumentSnapshot here too. This will probably be used with a useStaticCollection hook in the future.
+   */
+  endAt?: number
+  /**
+   * For now, this can only be a number, since it has to be JSON serializable.
+   *
+   * **TODO** allow DocumentSnapshot here too. This will probably be used with a useStaticCollection hook in the future.
+   */
+  startAfter?: number
+  /**
+   * For now, this can only be a number, since it has to be JSON serializable.
+   *
+   * **TODO** allow DocumentSnapshot here too. This will probably be used with a useStaticCollection hook in the future.
+   */
+  endBefore?: number
+
+  // THESE ARE NOT JSON SERIALIZABLE
```

Comment:
While we're at it, might be a good time to knock these out?

Reply:
Definitely yes! I'll look into that right now.

Reply:
Do you know if there is a way to construct a snapshot?

Reply:
I've managed to get it working with a snapshot startAt, etc. Instead of serializing and re-constructing the snapshot, I store it away in a map and then fetch it when deserializing.

I do not know how using multiple useCollection with the same snapshots will affect this. If it has problems, we could try changing the serializer from static to have one serializer per useCollection. Then, we can store the fuego inside the Serializer rather than passing it as a parameter, and we don't have to worry about other hooks interfering with the cache.

Reply:
@nandorojo Before merging, we need to investigate this above. Will it be a problem?